src/lib.rs

```diff
@@ -236,6 +237,24 @@ use prelude::*;
 
 const CONTENT_TYPE_JSON: &str = "application/json";
 
+#[derive(Clone)]
+/// This struct is used to maintain the strict URI encoding standard as proposed by RFC 3986
+struct StrictEncodeSet;
```

Comment:
https://github.com/rusoto/rusoto/blob/7640ca3de1e6486b0b25bce395810a9434a66ee5/rusoto/core/src/signature.rs#L593

Reply:
I'm finding the chain of specs referenced by https://tools.ietf.org/html/rfc6749#appendix-B to be a bit hard to follow, so it's not super clear to me exactly which characters need to be escaped (especially since we're dealing with an HTTP header here, rather than a URI). This code does seem to maintain the current behavior as implemented in https://github.com/curl/curl/blob/master/lib/escape.c#L40, but I wonder if it would suffice to use  
[`USERINFO_ENCODE_SET`](https://docs.rs/percent-encoding/1.0.1/percent_encoding/struct.USERINFO_ENCODE_SET.html), which seems to be intended for username and password escaping (according to https://github.com/dtolnay/url/blob/master/UPGRADING.md).

@DeeUnderscore do you have any thoughts here, as the author of #41? (see usage below)

Reply:
By my reading, this adheres to `application/x-www-form-urlencoded`—everything outside of the permitted bytes is encoded. [This WHATWG spec](https://url.spec.whatwg.org/#application/x-www-form-urlencoded) seems to be essentially equivalent to what RFC6749 wanted when it described the encoding in Appendix B. 

URLs can use forms of percent encoding which aren't `application/x-www-form-urlencoded`, since they permit more byte ranges (presumably to make URLs more readable). `USERINFO_ENCODE_SET` ([WHATWG ref](https://url.spec.whatwg.org/#userinfo-percent-encode-set)), in particular, is intended for the userinfo component of a URL, ie. the `user:password` in `https://user:password@example.org/`.

In practical terms, you could probably get away with a less strict encoding, and I suspect most OAuth implementations in the wild avoid using characters which could cause problems anyway. But, for closest adherence to the standard I'd recommend what this PR is doing.

Reply:
Thanks for the reference and rationale! That makes sense; I agree with the recommendation.

The encoding here does seem to deviate ever so slightly from the WHATWG spec, though:
- spaces encoded with `%20` instead of `+`
- asterisks (`0x2A`) excluded from the encoding set but mentioned by WHATWG
- tildes (`0x7E`) included in encoding set but excluded by WHATWG

None of these deviations seems particularly problematic, though it's probably safer for interoperability to err on the side of percent-encoding more values rather than fewer. Unless there's an error in the WHATWG spec, we should probably remove tildes from this set.